content/articles/2019-02-purrr-0-3-0.Rmd

```diff
@@ -0,0 +1,308 @@
+
+Other small improvements in consistency of behaviour are:
+
+* `modify_depth()` now operates at the atomic level as well.
+* `map()` and `modify()` now work with calls and pairlists.
+
+
+## Consistency of the interface
+
+### Direction of application
+
+The direction of application is now specified the same way across purrr functions. `reduce()`, `compose()` and `detect()` now have a `.dir` parameter that can take the value `"forward"` or `"backward"`. This terminology should be less ambiguous than "left" and "right":
+
+```{r, eval = FALSE}
+reduce(1:4, `-`, .dir = "backward")
+
+compose(foo, bar, .dir = "forward")
+
+detect(1:5, ~ . %% 2 == 0, .dir = "backward")
+```
+
+Note that the backward version of `reduce()` (called right-reduce in the literature) applies the reduced function in a slightly different way than `reduce_right()`. The new algorithm is more consistent with how this operation is usually defined in other languages.
+
+Following the introduction of the `.dir` parameters, the `_right` variants such as `reduce_right()` have been soft-deprecated, as well as the `.right` parameter of `detect()` and `detect_index()`.
+
+
+### partial()
+
+`partial()` has been rewritten to be a simple wrapper around `call_modify()` and `eval_tidy()` from the rlang package. Consequently, the `.env`, `.lazy` and `.first` arguments are soft-deprecated and replaced by a flexible syntax.
+
+To control the timing of evaluation, unquote the partialised arguments that should be evaluated only once when the function is created. The non-unquoted arguments are evaluated at each invokation of the function:
+
+```{r}
+my_list <- partial(list, lazy = rnorm(3), eager = !!rnorm(3))
+
+my_list()
+
+my_list()
+```
+
+You can also control the position of the future arguments by passing an empty `... = ` parameter. This syntax is powered by `rlang::call_modify()` to control where the position of dots in a function call:
+
+```{r}
+my_list <- partial(list, 1, ... = , 2)
+
+my_list()
+
+my_list("foo")
+```
+
+
+### `exec()` replaces `invoke()`
+
+We are retiring `invoke()` and `invoke_map()` in favour of `exec()`. Retirement means that we'll keep these functions indefinitely in the package, but we won't add features or recommend using them.
+
+We are now favouring `exec()`, which uses the tidy dots syntax for passing lists of arguments:
+
+```{r}
+# Before:
+invoke(mean, list(na.rm = TRUE), x = 1:10)
+
+# After
+exec(mean, 1:10, !!!list(na.rm = TRUE))
+```
+
+
+### Filling the missing parts
+
+* purrr 0.3.0 introduces `accumulate2()`, `modify2()` and `imodify()` variants.
+
+* By popular request, `at_depth()` is back as `map_depth()`. Unlike `modify_depth()` which preserves the class structure of the input tree, this variant only returns trees made of lists of lists, coercing vectors if needed.
```

Comment:
Maybe move this up to next to `map_at()` and have a "map tweaks" level-2 header?

Reply:
I added a Map tweaks section but I think it's better to keep `map_depth()` here, so we can keep the subsection "Filling the missing parts".